cpp/include/graph.hpp

```diff
@@ -214,5 +214,182 @@ class GraphCSC: public GraphCompressedSparseBase<VT,ET,WT> {
   {}
 };
 
+/**
+ * @brief       Base class for graphs constructed in the C++ API
+ *
+ * This class will own edge data, until the data is moved.
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class ConstructedGraphBase {
+public:
+  std::unique_ptr<WT>  edge_data{};     ///< edge weight
+  VT                   number_of_vertices{0};
+  ET                   number_of_edges{0};
+
+  ConstructedGraphBase() {}
+
+  ConstructedGraphBase(std::unique_ptr<WT> &&edge_data_, VT number_of_vertices_, ET number_of_edges_):
+    edge_data(edge_data_),
+    number_of_vertices(number_of_vertices_),
+    number_of_edges(number_of_edges_)
+  {}
+};
+
+/**
+ * @brief       A constructed graph stored in COO (COOrdinate) format.
+ *
+ * This class will src_indices and dst_indicies (until moved)
+ *
+ * @tparam VT   Type of vertex id
+ * @tparam ET   Type of edge id
+ * @tparam WT   Type of weight
+ */
+template <typename VT, typename ET, typename WT>
+class ConstructedGraphCOO: public ConstructedGraphBase<VT, ET, WT> {
```

Comment:
This class would allow k_core and k_truss to return a subgraph in COO format.  We can generate unique_ptr<VT> within the algorithms which will clean up if the algorithm aborts early.  At the end we pass them into the constructor for ConstructedGraphCOO, copy them into the output and return.